test/generate_parse_fixture_yml.py

```diff
@@ -1,34 +1,40 @@
 """Utility to generate yml files for all the parsing examples."""
+import multiprocessing
 import os
 
 import oyaml as yaml
 
-from sqlfluff.core.parser import Parser, Lexer
-from sqlfluff.core import FluffConfig
-from sqlfluff.cli.commands import quoted_presenter
+from conftest import compute_parse_tree_hash, parse_example_file
+from dialects.parse_fixtures import get_parse_fixtures
 
-from dialects.parse_fixtures import get_parse_fixtures, load_file
 
-yaml.add_representer(str, quoted_presenter)
-
-parse_success_examples, _ = get_parse_fixtures()
-
-for example in parse_success_examples:
+def generate_parse_fixture(example):
+    """Parse example SQL file, write parse tree to YAML file."""
     dialect, sqlfile = example
-    config = FluffConfig(overrides=dict(dialect=dialect))
-    # Load the SQL
-    raw = load_file(dialect, sqlfile)
-    # Lex and parse the file
-    tokens, _ = Lexer(config=config).lex(raw)
-    tree = Parser(config=config).parse(tokens)
-    r = None
-    if tree:
-        r = tree.as_record(code_only=True, show_raw=True)
+    tree = parse_example_file(dialect, sqlfile)
+    _hash = compute_parse_tree_hash(tree)
     # Remove the .sql file extension
     root = sqlfile[:-4]
     path = os.path.join("test", "fixtures", "parser", dialect, root + ".yml")
     with open(path, "w", newline="\n") as f:
-        if r:
+        r = None
+        if tree:
+            r = dict(
+                [("_hash", _hash)]
+                + list(tree.as_record(code_only=True, show_raw=True).items())
+            )
             yaml.dump(r, f, default_flow_style=False)
         else:
             f.write("")
+
+
+def main():
+    """Find all example SQL files, parse and create YAML files."""
+    parse_success_examples, _ = get_parse_fixtures()
+    with multiprocessing.Pool(multiprocessing.cpu_count()) as pool:
```

Comment:
Introduced multiprocessing so this script can process the `.sql` files more quickly. On my Mac, this change results in a ~3.5x speedup.

Reply:
Given this, are you 100% sure this PR is still the right approach?

With this speed up we could add this command to the tox.ini file with less of an impact and add the GitHub action as a fall back check just in case any contributor didn’t run it first.

This PR seems a lot of effort and change. If you still think it’s a good change then will bow to your better knowledge here but thought I’d just ask the question in case yours down the rabbit hole here.

Reply:
I'm not concerned about the amount of change -- it's mostly just moving code around with slight changes in ways that improve the maintainability. (TBH, I think the `test/generate_parse_fixture_yml.py` script began with copy-pasted code from other places, and this PR cleans that up a bit.) 

Can you remind me what the `tox` change would look like and how it would impact contributors' workflow? My thinking (possibly wrong) is that it would change the workflow for people who are already familiar with it. I actually don't use `tox` at all for SQLFluff development, but I don't know if that's common. I have manually created a virtualenv using `pyenv`. I use that for all my local development and testing, and I rely on the GitHub builds to do all the other testing (different Python versions, etc.)

Reply:
BTW, this PR is**tiny** compared to some of mine or Alan's. 😅 One reason I contribute to this project is because I often don't write much code in my real job. (I also contribute because it's useful to my real job. Kind of a win-win for me and my employer.)

Reply:
So my workflow is the following:

- Generate a virtualenv (only needs to be done once when first setting up the repo)
- Activate that virtual env (needs to be done each time you open a new shell):
```
source .venv/bin/activate
```
-  Create branch
- Make code change
- Generate YAML files (this is the step I missed in past as did them by hand - it also takes a minute or two, but maybe not with above change?):
```
python test/generate_parse_fixture_yml.py
```
- Run the quicker dialect tests only (I only discovered this command recently as hidden in a deeper README):
```
pytest test/dialects/dialects_test.py
```
- When fully finished, run the full tests for one python env only (takes a good few mins, even for one python version, so only really wanna run this when pretty much done):
```
tox -e cov-init,py38,cov-report,linting
```
- Check file changed, and config you're happy with the changes, and then commit them.
- Push to GitHub.
- Open PR.
- GitHub will automatically run all tests for ALL supported python versions in parallel.

I would propose the one line config change to `tox` as shown in https://github.com/sqlfluff/sqlfluff/issues/1277#issuecomment-898891659 to generate the YAML files when doing the full `tox` run. If people use `tox` before opening PR (like I do in above workflow) then they will automatically get the updated YAML files when they run it as so they will be included in the PR. If they don't, then that's fine too. And people can continue to run the generate command themselves if they want.

Either way, when they submit the PR a new GitHub Action will check the YAML files are up to date so if a contributor forgot to run it (or didn't know they had to), and it would produce any changes if it was run, then they will be informed and can run it manually and add to PR. If they always run it as part of their workflow (either manually, or via `tox`) then this check will never fail so is not additional effort. This would be a simple GitHub Action workflow to checkout the code, setup the python env for one python version (e.g. 3.8), run the YAML generation command, then use `git status` to see if the branch has any uncommitted changes and, if so, raise an alert. I'm happy to create this workflow as quite a lot of experience with GitHub Actions.

Does that make sense? Happy to answer any questions of create a sample PR to show you.

Reply:
> One reason I contribute to this project is because I often don't write much code in my real job. (I also contribute because it's useful to my real job. Kind of a win-win for me and my employer.)

Oh I hear you there! Same reasons I do open source in my spare time 😀

Reply:
Btw what would your workflow be with your change?

Is it that running `pytest` (either directly or via `tox`) fails if the YAML generation would have produced changes (i.e. you didn’t run it)? And so that would be caught by the existing checks on PRs and also if anyone runs `pytest` offline? If so, then that’s a good workflow too (maybe even a better one!) and works for me too. Just saying if that involves a lot of change and is much more complicated to implement, then there is the easier option of my proposal.

But there may be additional benefits to your change and would work better for those not using `tox` but running `pytest` directly.

Reply:
Yes, I'm pretty sure (but not 100% sure) that running pytest without running the script catches me if I forget to run the script.

IIUC, adding this step to the tox config would make it mandatory to also add the GitHub Action check as well, to ensure the files are up to date. If I'm correct that pytest already detects this, I don't see how this additional work will add any benefit.

Would you be willing to test the pytest approach to try and confirm or deny that it detects problems? The most common issues are:
* SQL file with no YAML file
* SQL and YAML files out of sync

Less common, but possible:
* YAML file with no SQL file
* Am I overlooking any others?

We may still decide to add the tox and GitHub steps, but having this background info will help us decide. I'd like to hear from @NiallRees  before we merge this change, since he was the original champion of adding YAML checks for every SQL file.

Reply:
> Yes, I'm pretty sure (but not 100% sure) that running pytest without running the script catches me if I forget to run the script.

If it does that already, then I'm confused what the point of this PR is?

> IIUC, adding this step to the tox config would make it mandatory to also add the GitHub Action check as well, to ensure the files are up to date.

That is correct. The `tox` additions just make it less likely to be a problem for those that run that.

> If I'm correct that pytest already detects this, I don't see how this additional work will add any benefit.
> Would you be willing to test the pytest approach to try and confirm or deny that it detects problems?

It doesn't catch minor errors and differences (e.g. using `'a'` over `"a"`) or spacing differences. Otherwise the extra changes in https://github.com/sqlfluff/sqlfluff/pull/1241 wouldn't have been caught?

It does capture when you handcraft YAML files and get it wrong. So thats good. But I didn't know to run that command so thought it was a bug. Hence why I raised #1273 

So the additional GitHub Action would capture those and have the following benefits:
1) Make contributors more aware of the auto generation.
2) Check PRs are including the correct YAML (as would be auto generated).
3) Prevent these randoms changes when people don't run this script before submitting a PR.
And, as I say, there's no real downside if people are running it already.

Reply:
I'm good with the tox and GitHub changes, too. Would you like to create a PR for my PR to add those changes?

Reply:
See https://github.com/sqlfluff/sqlfluff/pull/1280

Reply:
So, given that, do you still want to go the Hash route here? Not sure if that's overkill now or there's still other good changes and clean up in here?

We should at least do the parallelisation of the generation though as that's definitely a good change.

Oh and just realised you meant to add the PR to this PR - sorry I did a separate one!

Reply:
Definitely the parallel part. The hash stuff and the related code cleanup seem "nice to have", definitely not "must have". Let's see what Niall or other maintainers think. (I can't merge my own changes without someone else's review, regardless.)

Reply:
That's fine about your PR. I thought it may somehow depend on or touch the same code as this one, but they were completely independent. It worked out well -- I was happy to go ahead and get your changes in place! Good ideas there.